library/Fission/Internal/Text.hs

```diff
@@ -0,0 +1,18 @@
+module Fission.Internal.Text
+  ( split
+  , split'
+  ) where
+
+import           Fission.Prelude
+
+import           RIO.NonEmpty
+import qualified RIO.Text as Text
+
+
+split :: Int -> Text -> Maybe (NonEmpty Text)
+split len txt = nonEmpty $ Text.chunksOf len txt
+
+split' :: Int -> Text -> NonEmpty Text
+split' len txt = 
+  split len txt
+  |> maybe (pure "") identity
```

Comment:
Much more idiomatic: 

```haskell
split’ :: Natural -> Text -> NonEmpty Text
split’ len txt = fromMaybe (pure “”) $ split len txt
```

--- 

Is `””` a valid value for this use case?

Reply:
> Is ”” a valid value for this use case?

1) we'll actually be ditching this method in favor of `splitRecords` from the `ordered-txt` branch
2) kind of.... I actually went back and forth on how to handle it. It's a state that shouldn't ever be reached (unless you create a CID/Key wrapped around an empty string which would be silly). An empty string wrapped in a `NonEmpty` won't cause an error tho, it'll just set the dns TXT record to an empty string. So it's not an expected situation, should never happen, but still technically valid

Reply:
> So it's not an expected situation, should never happen, but still technically valid

Ideally we make that state unrepresentable, unless it’s a HUGE pain to do so. Basically use the type system to get rid of edge cases.

One simple way to do this is to use a `newtype`, and a “smart constructor” — a function like `nonEmpty` that does whatever runtime checks it needs. You make sure to not export the actual regular constructor, so the user can’t stick other values in there. Then we know that the value, once constructed, conforms to our exectations.

Reply:
I can walk you through that in more detail if helpful. It’s a pretty common pattern.

Reply:
yeah you've brought up "smart constructors" a couple times, so a walk through (or point to a good example) of that would probably be helpful

We'd want to do these checks on for CIDs & Keys

Reply:
> smart constructor

Sure thing! Here’s an example that’s actually possible to enforce the “regular” typed way, but is a huge pain to write that way: even numbers.

```haskell
module Even
  ( Even ()
  , toInt
  , mkEven
  ) where
newtype Even = Even {toInt :: Int}
  deriving newtype (Eq, Ord, Show, Display, Integral)
  
mkEven :: Integral a => a -> Maybe Even
mkEven a =
  if a `mod` 2 == 0
    then Just (Even a)
    else Nothing
```

The interesting part here is the exports. We don’t export the `Even` constructor . Normally you’d write the export line as `Even (..)`, but we want to keep that private to this module. No one can call it from outside.

In its place, we have `mkEven`, which does a runtime check that it is in fact even. If you’re able to get back an `Even` from that function, we now have a typesafe way of saying “this is an even number” without ever having to do the numerical check. `newtype`s get erased at compile time, and thus have zero runtime overhead.

The upshot here is that we only ever have to do the check once, and can have functions make the assumption that they have an even value. We can make this work with any arbitrary check (e.g. “is this email properly formatted?”). In this case, it would exclude the empty string case, and we don’t have to deal with that downstream, handle the errors of trying to pass an empty string to IPFS over the network, and so on. We restrict the possible space of errors by one, and have the compiler statically enforce an assumption about our code for us (it’s a”proof”).

Reply:
Also perhaps worth mentioning the anatomy of a datatype

```haskell
data TypeName = Constructor {field :: InnerType}
```

The left side (the part still getting exported) is the type (the thing after a `::`). The right side is the constructor (which you call as a value). So, by not exporting the constructor directly, users can depend on `TypeName`, but not make one with `Constructor`. They’re often the same name (“punning”) because they’re in different namespaces.

Sorry if that was review, but wanted to make sure 👍